test/e2e/scheduling/taint_based_evictions.go

```diff
@@ -0,0 +1,188 @@
+		zero := int64(0)
+		// build pod0, pod1
+		for i := 0; i < numPods-1; i++ {
+			pods[i] = createPausePod(f, pausePodConfig{
+				Name:     fmt.Sprintf("%v-%v", base, i),
+				NodeName: nodeName,
+				Tolerations: []v1.Toleration{
+					{
+						Key:               schedulerapi.TaintNodeUnreachable,
+						Operator:          v1.TolerationOpExists,
+						Effect:            v1.TaintEffectNoExecute,
+						TolerationSeconds: &tolerationSeconds[i],
+					},
+				},
+				DeletionGracePeriodSeconds: &zero,
+			})
+		}
+		// build pod2
+		pods[numPods-1] = createPausePod(f, pausePodConfig{
+			Name:     fmt.Sprintf("%v-%v", base, numPods-1),
+			NodeName: nodeName,
+		})
+
+		By("Verifying all pods are running properly")
+		for _, pod := range pods {
+			framework.ExpectNoError(framework.WaitForPodRunningInNamespace(cs, pod))
+		}
+
+		// get the node API object
+		nodeSelector := fields.OneTermEqualSelector("metadata.name", nodeName)
+		nodeList, err := cs.CoreV1().Nodes().List(metav1.ListOptions{FieldSelector: nodeSelector.String()})
+		if err != nil || len(nodeList.Items) != 1 {
+			framework.Failf("expected no err, got %v; expected len(nodes) = 1, got %v", err, len(nodeList.Items))
+		}
+		node := nodeList.Items[0]
+
+		By(fmt.Sprintf("Blocking traffic from node %s to the master", nodeName))
+		host, err := framework.GetNodeExternalIP(&node)
+		// TODO(Huang-Wei): make this case work for local provider
+		// if err != nil {
+		// 	host, err = framework.GetNodeInternalIP(&node)
+		// }
+		framework.ExpectNoError(err)
+		master := framework.GetMasterAddress(cs)
+		taint := newUnreachableNoExecuteTaint()
+
+		defer func() {
+			By(fmt.Sprintf("Unblocking traffic from node %s to the master", node.Name))
+			framework.UnblockNetwork(host, master)
+
+			if CurrentGinkgoTestDescription().Failed {
+				return
+			}
+
+			By(fmt.Sprintf("Expecting to see node %q becomes Ready", nodeName))
+			framework.WaitForNodeToBeReady(cs, nodeName, time.Minute*1)
+			By("Expecting to see unreachable=:NoExecute taint is taken off")
+			err := framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, false, time.Second*30)
+			framework.ExpectNoError(err)
+		}()
+
+		framework.BlockNetwork(host, master)
+
+		By(fmt.Sprintf("Expecting to see node %q becomes NotReady", nodeName))
+		if !framework.WaitForNodeToBeNotReady(cs, nodeName, time.Minute*3) {
+			framework.Failf("node %q doesn't turn to NotReady after 3 minutes", nodeName)
+		}
+		By("Expecting to see unreachable=:NoExecute taint is applied")
+		err = framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, true, time.Second*30)
+		framework.ExpectNoError(err)
+
+		By("Expecting pod0 to be evicted immediately")
+		err = framework.WaitForPodCondition(cs, ns, pods[0].Name, "pod0 terminating", time.Second*5, func(pod *v1.Pod) (bool, error) {
+			// as node is unreachable, pod0 is expected to be in Terminating status
+			// rather than getting deleted
+			if pod.DeletionTimestamp != nil {
```

Comment:
Shouldn't we check the pod status here?

Reply:
yeap, although `if pod.DeletionTimestamp != nil` is logically accurate, but adding a ANDed pod status check would be easier to understand.

Reply:
It looks like "Terminating" is more a _logic_ concept - say there is no API field indicating the pod is terminating, see:

https://github.com/kubernetes/kubernetes/blob/85860e28b28c6c30f6e189218e198b431c919742/pkg/printers/internalversion/printers.go#L659-L663

@ravisantoshgudimetla so are you fine with no change here?

Reply:
Yes. Thanks for checking.